csharp_dotnetcore/14.NLP-With-Dispatch/NlpDispatch/NlpDispatchBot.cs

```diff
@@ -0,0 +1,182 @@
+        /// In the .bot file, multiple instances of QnaMaker can be configured.
+        /// </summary>
+        public static readonly string QnAMakerKey = "sample.qna";
+
+        /// <summary>
+        /// Services configured from the ".bot" file.
+        /// </summary>
+        private readonly BotServices _services;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NlpDispatchBot"/> class.
+        /// </summary>
+        /// <param name="services">Services configured from the ".bot" file.</param>
+        public NlpDispatchBot(BotServices services)
+        {
+            _services = services ?? throw new System.ArgumentNullException(nameof(services));
+
+            if (!_services.QnAServices.ContainsKey(QnAMakerKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a QnA service named '{DispatchKey}'.");
+            }
+
+            if (!_services.LuisServices.ContainsKey(HomeAutomationLuisKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a Luis service named '{HomeAutomationLuisKey}'.");
+            }
+
+            if (!_services.LuisServices.ContainsKey(WeatherLuisKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a Luis service named '{WeatherLuisKey}'.");
+            }
+        }
+
+        /// <summary>
+        /// Every Conversation turn for our NLP Dispatch Bot will call this method.
+        /// There are no dialogs used, since it's "single turn" processing, meaning a single
+        /// request and response, with no stateful conversation.
+        /// </summary>
+        /// <param name="context">A <see cref="ITurnContext"/> containing all the data needed
+        /// for processing this conversation turn. </param>
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task OnTurnAsync(ITurnContext context, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (context.Activity.Type == ActivityTypes.Message && !context.Responded)
+            {
+                // Get the intent recognition result
+                var recognizerResult = await _services.LuisServices[DispatchKey].RecognizeAsync(context, cancellationToken);
+                var topIntent = recognizerResult?.GetTopScoringIntent();
+
+                if (topIntent == null)
+                {
+                    await context.SendActivityAsync("Unable to get the top intent.");
+                }
+                else
+                {
+                    if (topIntent.Value.score < 0.3)
```

Comment:
> 0.3 [](start = 47, length = 4)

I have a very strong distrust of magic numbers. What is this? What does it mean? Why isn't it 0.4? Or 0.2? 

Links to docs here is the only option, unfortunatly.

Reply:
I have the same distrust too, completely agree. I had an email going to answer how to do this correctly, my bad on not mentioning it. I'll get this right next iteration.

Reply:
Resolved!